File: rust_study/rust_luogu/rust_codeforces/cpp/luogu/day3/p2024.cpp
// Copyright 2024 Allergy
// email     Allergy527@gmail.com
// workspace luogu\day3\p2024.cpp
// date      2024/09/24 14:14:43
#include <bits/stdc++.h>
// #define int int64_t

using namespace std;  // NOLINT
using ll = int64_t;
int inf = 0x3f3f3f3f;
ll ll_inf = 0x3f3f3f3f3f3f3f3f;

struct UFSet {
    vector<int> fa, rel;
    UFSet(int n) : fa(n + 1), rel(n + 1) {  // NOLINT
        iota(fa.begin(), fa.end(), 0);
    }
    int find(int x) {
        if (fa[x] == x) return x;
        int temp = find(fa[x]);
        rel[x] = (rel[x] + rel[fa[x]]) % 3;
        return fa[x] = temp;
    }
    void merge(int x, int y, int rootx, int rooty, int f) {
        fa[rootx] = rooty;
        rel[rootx] = (3 - rel[x] + rel[y] + f - 1) % 3;
    }
};

void solve() {
    int n, k, ans = 0;
    cin >> n >> k;
    UFSet qwq(n);
    for (int i = 0; i < k; i++) {
        int f, a, b;
        cin >> f >> a >> b;
        if (a > n || b > n || (f == 2 && a == b)) {
            ++ans;
            continue;
        }
        int rootx = qwq.find(a);
        int rooty = qwq.find(b);
        if (f == 1) {
            if (rootx == rooty && qwq.rel[a] != qwq.rel[b]) {
                ++ans;
                continue;
            } else if (rootx != rooty) {
                qwq.merge(a, b, rootx, rooty, f);
            }
        } else {
            if (rootx == rooty) {
                if ((qwq.rel[a] - qwq.rel[b] + 3) % 3 != 1) {
                    ++ans;
                    continue;
                }
            } else {
                qwq.merge(a, b, rootx, rooty, f);
            }
        }
    }
    cout << ans << '\n';
}

signed main() {
    ios::sync_with_stdio(false);
    int t = 1;
    // cin >> t;
    while (t--) solve();
    cin >> t;
    return 0;
}

File: rust_luogu/src/bin/p2024.rs
//@author    Allergy
//@email     Allergy527@gmail.com
//@workspace Prectice_Contest/p2024.rs
//@data      2024/10/22 21:28:33
fn cin() -> String {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    input.trim().to_string()
}
fn main() {
    let t = 1; //let t = cin().parse::<i32>().unwrap();
    (0..t).for_each(|_| solve());
}
struct UFset {
    fa: Vec<usize>,
    rel: Vec<usize>,
}
impl UFset {
    fn new(n: usize) -> Self {
        Self {
            fa: (0..=n).collect::<Vec<_>>(),
            rel: vec![0; n + 1],
        }
    }
    fn find(&mut self, x: usize) -> usize {
        if x != self.fa[x] {
            let tmp = self.fa[x];
            self.fa[x] = self.find(self.fa[x]);
            self.rel[x] = (self.rel[x] + self.rel[tmp]) % 3;
        }
        self.fa[x]
    }
    fn merge(&mut self, x: usize, y: usize, f: usize) {
        let (fx, fy) = (self.find(x), self.find(y));
        self.fa[fx] = fy;
        self.rel[fx] = (3 - self.rel[x] + self.rel[y] + f - 1) % 3; //f为1,同类，f为2，x吃y
    }
}

fn solve() {
    let (n, k) = cin()
        .split_whitespace()
        .fold((0, 0), |x, y| (x.1, y.parse::<usize>().unwrap()));
    let mut qwq = UFset::new(n);
    let ans = (0..k).fold(0, |ans, _| {
        let (f, a, b) = cin()
            .split_whitespace()
            .fold((0, 0, 0), |x, y| (x.1, x.2, y.parse::<usize>().unwrap()));
        if a > n || b > n {
            return ans + 1;
        }
        //放进来了
        if qwq.find(a) == qwq.find(b) {
            //b吃a或同类
            if qwq.rel[a] != (qwq.rel[b] + f - 1) % 3 {
                return ans + 1;
            }
        } else {
            qwq.merge(a, b, f);
        }
        ans
    });
    println!("{}", ans);
}
